VirtoCommerce.CatalogModule.Data/Services/ProductAssociationSearchService.cs

```diff
@@ -0,0 +1,131 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using VirtoCommerce.CatalogModule.Data.Repositories;
+using VirtoCommerce.Domain.Catalog.Model;
+using VirtoCommerce.Domain.Catalog.Model.Search;
+using VirtoCommerce.Domain.Catalog.Services;
+using VirtoCommerce.Domain.Commerce.Model.Search;
+using VirtoCommerce.Domain.Search;
+using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.Platform.Data.Infrastructure;
+
+namespace VirtoCommerce.CatalogModule.Data.Services
+{
+  public class ProductAssociationSearchService : IProductAssociationSearchService
+  {
+    private readonly Func<ICatalogRepository> _catalogRepositoryFactory;
+    private readonly IItemService _itemService;
+    private readonly ICategoryService _categoryService;
+
+    private readonly Dictionary<string, string> _productSortingAliases = new Dictionary<string, string>();
+
+    public ProductAssociationSearchService(Func<ICatalogRepository> catalogRepositoryFactory, IItemService itemService, ICategoryService categoryService)
+    {
+      _catalogRepositoryFactory = catalogRepositoryFactory;
+      _itemService = itemService;
+      _categoryService = categoryService;
+
+      _productSortingAliases["sku"] = ReflectionUtility.GetPropertyName<CatalogProduct>(x => x.Code);
+    }
+
+    public GenericSearchResult<CatalogProduct> SearchProductAssociations(ProductAssociationSearchCriteria criteria)
+    {
+      var retVal = new GenericSearchResult<CatalogProduct>();
+   
+      var sortInfos = criteria.SortInfos;
+      if (sortInfos.IsNullOrEmpty())
+      {
+        sortInfos = new[] { new SortInfo { SortColumn = "Priority", SortDirection = SortDirection.Descending }, new SortInfo { SortColumn = "Name", SortDirection = SortDirection.Ascending } };
+      }
+      //Try to replace sorting columns names
+      TryTransformSortingInfoColumnNames(_productSortingAliases, sortInfos);
+   
+      var productResponseGroup = ItemResponseGroup.ItemInfo | ItemResponseGroup.ItemAssets | ItemResponseGroup.Links | ItemResponseGroup.Seo;
+   
+      var objectTypes = criteria.ObjectTypes.ToList();
+      var objectIds = criteria.ObjectIds;
+      var searchCategoryIds = new string[] { };
+      var categoryIds = new string[] { };//new List<string>();
+   
+      var products = _itemService.GetByIds(objectIds.ToArray(), productResponseGroup)
+                              .OrderBy(x => objectIds.IndexOf(x.Id)).ToList();   
+    
+      using (var repository = _catalogRepositoryFactory())
+      {
+        //Optimize performance and CPU usage
+        repository.DisableChangesTracking();
+       
+        foreach (var product in products)
+        {
+          if (product.Associations.Count() > 0)
+          {
+             for(var i = 0; i < product.Associations.ToArray().Length; i++)
+             {
+               categoryIds[i] = product.CategoryId;
+             }
+          }
+          //if (product.ReferencedAssociations.Count() > 0)
```

Comment:
remove this comment

Reply:
Remove all this code
and use instead
ICatalogSearchService.Search( new SearchCriteria(ResponseGroup = ResponseGroup.WithItems, SearchInChildren = true, CategoryIds = categoryAssociationsIds)